feat: Added kinesis exporter and respective tests.

baitcode opened the pull request:
## Description
Added kinesis exporter and respective tests.

I'm finalising the issue. Need to update documentation and perform manual testing and add an example.

## Closes issue(s)
Resolve #2557

## Checklist
- [x] I have tested this coded
- [x] I have added unit test to cover this code
- [x] I have updated the documentation (`README.md` and `/website/docs`)
- [x] I have followed the [contributing guide](CONTRIBUTING.md)

thomaspoignant commented on README.md:
```diff
@@ -553,6 +553,7 @@ It collects all variation events and can save these events in several locations:
 - **Local file** *- create local files with the variation usages.*
 - **Log** *- use your logger to write the variation usages.*
 - **AWS S3** *- export your variation usages to S3.*
+- **AWS S3** *- publish your variation usages to AWS Kinesis Stream.*
```
```suggestion
- **AWS Kinesis** *- publish your variation usages to AWS Kinesis Stream.*
```

thomaspoignant reviewed (changes requested):
@baitcode thanks a lot for this exporter, this looks really great.
I need to do a proper test using AWS to check that everything goes well but it sounds really good.

There is still a missing part, it is the possibility to configure it in the relay-proxy.
The way to do it is to manage the configuration in this [file](https://github.com/thomaspoignant/go-feature-flag/blob/main/cmd/relayproxy/config/exporter.go).
_If you need help for this I can help._ 

I will do a second review, after setting up a Kinesis on AWS and I let you know how it goes.

One more thing, it is about the comment in the issue about the sharding part, do you think we can let the user configure this when setting up the exporter? 
Ideally we should use the user targetingKey as a sharding key if possible.

baitcode commented:
@thomaspoignant 

I've updated PR. 

I could not figure out `targetingKey`, but, instead, I converted PartitionKey to a function calculating key based on FeatureEvent and Context object. I don't want to get to much deep into details of `targetingKey` implementation. Seems it's being passed through context, but it's goes through all the layers of the application and unfortunately I don't have a capacity to do that research neither I can find doc covering it. Please forget me for that, I would greatly appreciate if you relieved me from that responsibility. Thank you in advance.

baitcode commented:
@thomaspoignant do you mind I fix all the linters and Codecoverage?

thomaspoignant commented:
> @thomaspoignant do you mind I fix all the linters and Codecoverage?

No please fix it 🙏

thomaspoignant commented on cmd/relayproxy/service/gofeatureflag.go:
```diff
@@ -260,6 +261,19 @@ func createExporter(c *config.ExporterConf) (exporter.CommonExporter, error) {
 			ParquetCompressionCodec: parquetCompressionCodec,
 			AwsConfig:               &awsConfig,
 		}, nil
+	case config.KinesisExporter:
+		awsConfig, err := awsConf.LoadDefaultConfig(context.Background())
+		if err != nil {
+			return nil, err
+		}
+		return &kinesisexporter.Exporter{
+			Format:    format,
+			AwsConfig: &awsConfig,
+			Settings: kinesisexporter.NewSettings(
+				kinesisexporter.WithStreamArn(c.StreamArn),
+				kinesisexporter.WithStreamName(c.StreamName),
+			),
+		}, nil
```
We should maybe add a test for this one.

thomaspoignant commented on exporter/kinesisexporter/exporter.go:
```diff
@@ -0,0 +1,220 @@
+	Mb         = 1024 * 1024
+)
+
+var DefaultPartitionKey = func(event exporter.FeatureEvent) string {
```
Proposal to avoid any linter issue
```suggestion
var DefaultPartitionKey = func(_ exporter.FeatureEvent) string {
```

baitcode commented:
@thomaspoignant 
As all checks are locked again and I can't see the result of CI execution I used `make coverage` and `make lint` commands to get at least some data. Kinesis is now 85% covered. Added relay case to test list (didn't do much it's coverage is 50% roughly).
I guess that should do the work.

thomaspoignant commented:
> @thomaspoignant As all checks are locked again and I can't see the result of CI execution I used `make coverage` and `make lint` commands to get at least some data. Kinesis is now 85% covered. Added relay case to test list (didn't do much it's coverage is 50% roughly). I guess that should do the work.

No worries for that, 85% is really good.

thomaspoignant reviewed (approved):
We are good to go @baitcode.
Thanks a lot for your contribution, I will merge it and it will be part of the next version.
